src/jswrap_date.c

```diff
@@ -19,29 +19,157 @@
 #include "jslex.h"
 
 const int MSDAY = 24*60*60*1000;
-const int YDAY = 365;
-const int LDAY = 366;
-const int FDAY = 4*365+1;
 const int BASE_DOW = 4;
-const short DAYS[13] = {0,31,59,90,120,151,181,212,243,273,304,334,365};
-const short LPDAYS[13] = {0,31,60,91,121,152,182,213,244,274,305,335,366};
-const short YDAYS[4] = {0,365,365*2,365*3+1};
 const char *MONTHNAMES = "Jan\0Feb\0Mar\0Apr\0May\0Jun\0Jul\0Aug\0Sep\0Oct\0Nov\0Dec";
 const char *DAYNAMES = "Sun\0Mon\0Tue\0Wed\0Thu\0Fri\0Sat";
 
-/// return time zone in minutes
-int jsdGetTimeZone() {
+// Convert a y,m,d into a number of days since 1970. 0<=m<=11
+int getDayNumberFromDate(int y, int m, int d) {
+  int ans;
+
+  if (m < 2) {
+    y--;
+    m+=12;
+  }
+  ans = (y/100);
+  ans = 365*y + (y>>2) - ans + (ans>>2) + 30*m + ((3*m+6)/5) + d - 719531;
+  return ans;
+}
+
+// Convert a number of days since 1970 into y,m,d. 0<=m<=11
+void getDateFromDayNumber(int day, int *y, int *m, int *date) {
```

Comment:
Is there any kind of reference for this that could be linked? This looks a lot like witchcraft to me and it might just be handy to know where it came from.

Reply:
It is adopted from an algorithm in a book called "Astronomical Algorithms" by Jean Meeus. The book - from 1991 - is well out of print by now.

I'm in the process of doing [my own write-up on these algorithms](https://github.com/deirdreobyrne/CalendarAndDST). I've managed to convince myself of the correctness of everything but the constant 2877911. I've also ran getDayNumberFromDate against getDateFromDayNumber, and they are consistent.

Reply:
I had to do quite a bit of work to the algorithm in "Astronomical Algorithms", as it had its year zero in the year 4713BC. I wanted my year zero to be 1970 to correspond with the zero of Unix time, and the conversion was quite the mind bend.

I just *love* algorithms like that - it was a joy to work on!